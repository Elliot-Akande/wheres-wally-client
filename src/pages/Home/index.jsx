import { Link } from "react-router-dom";
import useFetch from "../../hooks/useFetch";
import styles from "./index.module.css";
import ImageLoader from "../../components/ImageLoader/ImageLoader";

const Home = () => {
  const { data, loading, error } = useFetch("/levels");

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error fetching level data.</p>;

  return (
    <div className={styles.container}>
      {data.map(({ levelNum, imageUrl }) => (
        <LevelCard levelNum={levelNum} imageUrl={imageUrl} key={levelNum} />
      ))}
    </div>
  );
};

const LevelCard = ({ levelNum, imageUrl }) => {
  return (
    <Link to={`level/${levelNum}/details`} className={styles.card}>
      <ImageLoader
        src={imageUrl}
        alt={`Level ${levelNum}`}
        imgClass={styles.img}
        placeholderClass={styles.placeholder}
        spinnerColour={"hsl(288, 70%, 49%)"}
      />
      <div className={styles.title}>LEVEL {levelNum}</div>
    </Link>
  );
};

export default Home;
