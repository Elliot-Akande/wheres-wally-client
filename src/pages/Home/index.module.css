.container {
  width: min(800px, 100%);
  padding: 48px 0;

  display: flex;
  flex-direction: column;
  gap: 24px;
}

.card {
  height: 300px;
  position: relative;

  border-radius: 16px;
  transition: 0.2s;
  overflow: hidden;
}

.card:hover {
  transform: scale(1.02);
}

.card::after {
  height: 200px;
  width: 200px;
  padding: 12px;

  content: "Play!";
  position: absolute;
  bottom: 0;
  right: 0;

  display: flex;
  justify-content: center;

  font-size: 24px;
  color: var(--neutral-100);
  background: linear-gradient(180deg, var(--accent), var(--accent-dark) 32%);

  opacity: 0;
  transform: translate(100%, 100%) rotate(-45deg);
  transition: all 0.2s;
}

.card:hover::after {
  opacity: 1;
  transform: translate(60%, 60%) rotate(-45deg);
}

.img {
  min-width: 100%;
  height: 100%;
  object-fit: cover;

  border-radius: 16px;
}

.title {
  padding: 16px 32px;

  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;

  font-weight: bold;
  font-size: 24px;
  color: var(--neutral-100);
  background: linear-gradient(180deg, var(--neutral-900), transparent 50%);
}
