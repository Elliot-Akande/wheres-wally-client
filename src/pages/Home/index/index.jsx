import useFetch from "../../../hooks/useFetch";
import LevelCard from "../LevelCard/LevelCard";
import { StyledHome } from "./styles";

const Home = () => {
  const { data, loading, error } = useFetch("/levels");

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error fetching level data.</p>;

  return (
    <StyledHome>
      {data.map(({ levelNum, imageUrl }) => (
        <LevelCard levelNum={levelNum} imageUrl={imageUrl} key={levelNum} />
      ))}
    </StyledHome>
  );
};

export default Home;
