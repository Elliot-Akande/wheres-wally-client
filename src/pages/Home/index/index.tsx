import useFetch from "../../../hooks/useFetch";
import LevelCard from "../LevelCard/LevelCard";
import { StyledHome } from "./styles";

type Data = Array<{
  levelNum: string;
  imageUrl: string;
}>;

const Home = () => {
  const { data, loading, error } = useFetch<Data>("/levels");

  if (loading) return <p>Loading...</p>;
  if (error || !data) return <p>Error fetching level data.</p>;

  return (
    <StyledHome>
      {data.map(({ levelNum, imageUrl }) => (
        <LevelCard levelNum={levelNum} imageUrl={imageUrl} key={levelNum} />
      ))}
    </StyledHome>
  );
};

export default Home;
