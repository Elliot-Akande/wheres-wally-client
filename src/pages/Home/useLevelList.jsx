import { useState, useEffect } from "react";

const useLevelList = () => {
  const [data, setData] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    setData([
      {
        levelNum: 1,
      },
      {
        levelNum: 2,
      },
      {
        levelNum: 3,
      },
      {
        levelNum: 4,
      },
    ]);
    setLoading(false);
  }, []);

  return { data, loading, error };
};

export default useLevelList;
