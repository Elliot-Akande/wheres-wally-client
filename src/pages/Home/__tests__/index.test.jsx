import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { describe, expect, it, vi } from "vitest";
import Home from "../index.jsx";
import useLevelList from "../useLevelList";

const mocks = vi.hoisted(() => ({
  useLevelList: vi.fn(),
}));

vi.mock("../useLevelList.jsx", () => ({
  default: mocks.useLevelList,
}));

const data = [
  {
    levelNum: 1,
  },
  {
    levelNum: 2,
  },
  {
    levelNum: 3,
  },
];
const loading = false;
const error = null;

describe("Home", () => {
  it("renders Level links", () => {
    mocks.useLevelList.mockReturnValue({
      data,
      loading,
      error,
    });

    const { container } = render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    expect(useLevelList).toBe(mocks.useLevelList);
    expect(useLevelList).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });

  it("renders loading message on useLevelList loading", () => {
    mocks.useLevelList.mockReturnValue({
      data,
      loading: true,
      error,
    });

    const { container } = render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    expect(useLevelList).toBe(mocks.useLevelList);
    expect(useLevelList).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });

  it("renders error message on useLevelList error", () => {
    mocks.useLevelList.mockReturnValue({
      data,
      loading,
      error: new Error(),
    });

    const { container } = render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    expect(useLevelList).toBe(mocks.useLevelList);
    expect(useLevelList).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });
});
