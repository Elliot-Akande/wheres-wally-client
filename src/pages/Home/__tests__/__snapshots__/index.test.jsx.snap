// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > renders Level links 1`] = `
<div>
  <a
    href="/level/1/details"
  >
    Level 
    1
  </a>
  <a
    href="/level/2/details"
  >
    Level 
    2
  </a>
  <a
    href="/level/3/details"
  >
    Level 
    3
  </a>
</div>
`;

exports[`Home > renders error message on useLevelList error 1`] = `
<div>
  <p>
    Data fetching error.
  </p>
</div>
`;

exports[`Home > renders loading message on useLevelList loading 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;
